File: frontend/iosApp/UILogics/Sources/UILogics/Screens/SyncQueuesViewModel.swift
import Foundation
import Combine
import Domains
import UseCases

@MainActor
public final class SyncQueuesViewModel: ObservableObject {
    @Published public private(set) var items: [SyncOperationUIModel] = []

    private let useCase: SyncQueuesUseCaseProtocol
    private var cancellables = Set<AnyCancellable>()

    public init(useCase: SyncQueuesUseCaseProtocol) {
        self.useCase = useCase

        useCase.observeState()
            .receive(on: DispatchQueue.main)
            .sink { [weak self] _ in
                Task {
                    await self?.loadOperations()
                }
            }
            .store(in: &cancellables)
    }

    public func onAppear() {
        Task {
            await loadOperations()
        }
    }

    private func loadOperations() async {
        let queueItems = await useCase.getAll()
        items = queueItems.map { SyncOperationUIModel(from: $0) }
    }
}

// MARK: - UI Models

extension SyncQueuesViewModel {
    public struct SyncOperationUIModel: Identifiable, Equatable {
        public let id: UUID
        public let entityType: String
        public let operationType: String
        public let status: Status
        public let errorMessage: String?
        public let entityTitle: String?
        public let localId: String
        public let serverId: String?
        public let createdAt: String

        public init(from item: SyncQueueItem) {
            let operation = item.operation
            self.id = operation.idUUID
            self.entityType = Self.mapEntityType(operation.entityType)
            self.operationType = Self.mapOperationType(operation.operationType)
            self.status = Self.mapStatus(operation.status)
            self.errorMessage = operation.errorMessage
            self.entityTitle = item.entityTitle
            self.localId = String(operation.localIdUUID.uuidString.prefix(8)).lowercased()
            self.serverId = item.entityServerId.map { String($0) }
            self.createdAt = Self.formatDate(operation.createdAtDate)
        }

        private static func mapEntityType(_ type: EntityType) -> String {
            switch type {
            case .album: return "Album"
            case .memory: return "Memory"
            case .user: return "User"
            }
        }

        private static func mapOperationType(_ type: OperationType) -> String {
            switch type {
            case .create: return "Create"
            case .update: return "Update"
            }
        }

        private static func mapStatus(_ status: SyncOperationStatus) -> Status {
            switch status {
            case .pending: return .pending
            case .inProgress: return .inProgress
            case .failed: return .failed
            }
        }

        private static func formatDate(_ date: Date) -> String {
            let formatter = DateFormatter()
            formatter.dateStyle = .short
            formatter.timeStyle = .medium
            return formatter.string(from: date)
        }

        public enum Status: Equatable {
            case pending
            case inProgress
            case failed

            public var displayName: String {
                switch self {
                case .pending: return "Pending"
                case .inProgress: return "In Progress"
                case .failed: return "Failed"
                }
            }
        }
    }
}

File: frontend/composeApp/src/androidMain/kotlin/com/example/memoriesapp/ui/uilogics/viewmodels/SyncQueuesViewModel.kt
package com.example.memoriesapp.ui.uilogics.viewmodels

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.memoriesapp.domain.SyncOperationStatus
import com.example.memoriesapp.usecase.SyncQueuesUseCase
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.launch
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

class SyncQueuesViewModel(
    private val syncQueuesUseCase: SyncQueuesUseCase
) : ViewModel() {

    var items by mutableStateOf<List<SyncOperationUIModel>>(emptyList())
        private set

    init {
        // Observe state changes and refresh list
        syncQueuesUseCase.observeState()
            .onEach { loadItems() }
            .launchIn(viewModelScope)
    }

    fun onAppear() {
        viewModelScope.launch {
            loadItems()
        }
    }

    private suspend fun loadItems() {
        val queueItems = syncQueuesUseCase.getAll()
        items = queueItems.map { item ->
            val op = item.operation
            SyncOperationUIModel(
                id = op.id.toString(),
                entityType = op.entityType.name,
                operationType = op.operationType.name,
                entityTitle = item.entityTitle,
                localId = op.localId.toString().take(8),
                serverId = item.entityServerId?.toString(),
                createdAt = formatDate(op.createdAt.epochMillis),
                status = mapStatus(op.status),
                errorMessage = op.errorMessage
            )
        }
    }

    private fun formatDate(timestamp: Long): String {
        val date = Date(timestamp)
        val format = SimpleDateFormat("yyyy-MM-dd HH:mm", Locale.getDefault())
        return format.format(date)
    }

    private fun mapStatus(status: SyncOperationStatus): SyncOperationUIModel.Status {
        return when (status) {
            SyncOperationStatus.PENDING -> SyncOperationUIModel.Status.PENDING
            SyncOperationStatus.IN_PROGRESS -> SyncOperationUIModel.Status.IN_PROGRESS
            SyncOperationStatus.FAILED -> SyncOperationUIModel.Status.FAILED
        }
    }

    data class SyncOperationUIModel(
        val id: String,
        val entityType: String,
        val operationType: String,
        val entityTitle: String?,
        val localId: String,
        val serverId: String?,
        val createdAt: String,
        val status: Status,
        val errorMessage: String?
    ) {
        enum class Status(val displayName: String) {
            PENDING("Pending"),
            IN_PROGRESS("In Progress"),
            FAILED("Failed")
        }
    }
}
